linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -227,25 +227,17 @@ private final class IRChecker(unit: LinkingUnit, reporter: ErrorReporter,
 
       case Assign(lhs, rhs) =>
         def checkNonStaticField(receiver: Tree, name: FieldName): Unit = {
-          val fieldClassName = name.className
-
-          def allowWriteImmutable = env.inConstructorOf match {
-            case Some(`fieldClassName`) =>
-              // ctors can write immutable fields of the class they are constructing.
-              true
-
-            case Some(ctorClassName) if postOptimizer =>
-              // postOptimizer, due to ctor inlining, we may write immutable parent class fields as well.
-              lookupClass(ctorClassName).ancestors.contains(fieldClassName)
-
-            case _ => false
-          }
-
           receiver match {
-            case This() if allowWriteImmutable =>
-              // ok
+            case This() if postOptimizer && env.inConstructorOf.isDefined ||
+                env.inConstructorOf == Some(name.className) =>
```

Comment:
For clarity, consider:
```suggestion
            case This() if (postOptimizer && env.inConstructorOf.isDefined) ||
                env.inConstructorOf == Some(name.className) =>
```